utensor_cgen/ir/base.py

```diff
@@ -0,0 +1,271 @@
+    return (hasattr(value, 'CopyFrom') or
+            type(value).__module__.startswith('google.protobuf'))
+
+
+@attr.s
+class TensorInfo(IRBase, _NoShallowCopyMixin):
+  """
+  name : str
+  dtype : numpy.dtype
+  shape : list
+  """
+  name = attr.ib(validator=validators.instance_of(six.text_type))
+  dtype = attr.ib(validator=validators.instance_of(np.dtype))
+  shape = attr.ib(validator=validators.instance_of((list, type(None))))
+  @shape.validator
+  def check(self, attrib, values):
+    if values is not None:
+      for v in values:
+        assert isinstance(v, (int, type(None))), \
+          "shape should be a list of integers"
+
+  def __deepcopy__(self, memo):
+    return TensorInfo(name=self.name,
+                      dtype=self.dtype,
+                      shape=deepcopy(self.shape, memo))
+
+  # legacy code: to make it works like namedtuple
+  def __iter__(self):
+    # TODO remove all such code in utensor
+    #   name, dtype, shape = tensor_info
+    return iter((self.name, self.dtype, self.shape))
+
+  def __getitem__(self, k):
+    return (self.name, self.dtype, self.shape)[k]
+
+
+@attr.s
+class OperationInfo(IRBase, _NoShallowCopyMixin):
+  """
+  name : str
+  input_tensors : List[TensorInfo]
+  output_tensors : List[TensorInfo]
+  op_type : str
+  backend : str
+  op_attr : dict
+  """
+  name = attr.ib(validator=validators.instance_of(six.text_type))
+
+  input_tensors = attr.ib(validator=validators.instance_of(list))
+  @input_tensors.validator
+  def check(self, attribute, value):
+    if not all([isinstance(v, TensorInfo) for v in value]):
+      raise ValueError('Expecting a list of TensorInfo for input_tensor')
+
+  output_tensors = attr.ib(validator=validators.instance_of(list))
+  @output_tensors.validator
+  def check(self, attribute, value):
+    if not all([isinstance(v, TensorInfo) for v in value]):
+      raise ValueError('Expecting a list of TensorInfo for output_tensor')
+
+  op_type = attr.ib(validator=validators.instance_of(six.text_type))
+
+  backend = attr.ib(validator=validators.instance_of(six.text_type))
+  @backend.validator
+  def check(self, attribute, value):
+    if value not in ['tensorflow']:
+      raise ValueError('Unsupported backend: {}'.format(value))
+
+  op_attr = attr.ib(factory=dict)
+
+  def __attrs_post_init__(self):
+    if self.op_attr:
+      op_attr = {}
+      for k, v in self.op_attr.items():
+        op_attr[k] = ConverterFactory.get_generic_value(v)
+      self.op_attr = op_attr
+  
+  def __deepcopy__(self, memo):
+    return OperationInfo(name=self.name,
+                         input_tensors=deepcopy(self.input_tensors, memo),
+                         output_tensors=deepcopy(self.output_tensors, memo),
+                         op_type=self.op_type,
+                         backend=self.backend,
+                         op_attr=deepcopy(self.op_attr, memo))
+
+
+class uTensorGraph(IRBase, _NoShallowCopyMixin):
+  """
+  Attributes
+  ==========
```

Comment:
I found this definition is not correct for a freezed graph.
Since all variables are also transformed into const tensors which are noes with no inputs, either.
This definition will mix everything up.

Make it required argument?

Reply:
hmm i think we should leave it for now then